Async filtering to /cards/ and /rotation/ based on Scryfall Queries

tsutton opened the pull request:
This pull request is built from parts of #6334 and improvements since then. It adds a form to /cards/ and /rotation/ (including /rotation/new/ etc) which lets the user input a query, then submits that to Scryfall and filters the page in-place. It supports browser history as well.

tsutton committed:
Filtering by scryfall queries in /cards/

tsutton committed:
Cleaner /card/ filtering with support for browser back and forward and error handling

tsutton committed:
Improved browser history handling for /cards/ filtering

tsutton committed:
Add new filter to /rotation/ and make filters hidden behind a button on load

tsutton committed:
Correct behavior of visiting /rotation/?fq= pages directly

tsutton committed:
Remove TODO

bakert commented on decksite/templates/cards.mustache:
```diff
@@ -1,4 +1,8 @@
 {{> banner}}
 <section>
+    <button id="toggle-filters-button" onclick="PD.filter.toggleDisplayFilter()">Show filters</button>
```
Please attach the onclick behavior in pd.js don’t inline it here. It’s our standard.

bakert commented on decksite/templates/cardstable.mustache:
```diff
@@ -6,23 +6,33 @@
-            <th class="n" data-breakpoints="xs" data-sortInitialOrder="desc">⑧</th>
-            <th class="n" data-breakpoints="xs" data-sortInitialOrder="desc">5–0s</th>
+            <th class="n" data-breakpoints="xs" data-sortInitialOrder="desc">
+                <abbr title="Tournament wins">①</abbr>
```
If we’re going to do this we should do it everywhere that we use these symbols.

bakert commented on decksite/main.py:
```diff
@@ -320,9 +323,10 @@ def community_guidelines() -> str:
-    view = Rotation(interestingness, rotation_query, cardnames)
+    query = request.args.get('fq')
+    if query is None:
+        query = ''
```
If it’s easy it would be good to add this to /rotation/changes and /rotation/speculation too. That can definitely be in a follow up PR tho.

tsutton replied:
Good idea, it shoudn't be hard. I'll make it in a follow up.

bakert commented on decksite/templates/cardstable.mustache:
```diff
@@ -6,23 +6,33 @@
     <tbody>
         {{#cards}}
-            <tr>
+            <tr class="cardrow" id="cardrow-{{name}}">
```
I’m not a big fan of ids in html in general. Things you expect to remain unique have a tendency to become ununique. In this case it’s pretty easy to imagine a page with two cardtables on it where the same card appears in both tables.  Can we make this a class?  I’d also sanitize the name down to lowercase letters and hyphens only personally but that’s less important. Hopefully gzip means this doesn’t really add much page weight to something like /seasons/all/cards but might be worth a spot check.

tsutton replied:
Changed to not using id. Everything is stuffed into classes, except I put the card name into a data attribute, which seemed like a fancy thing that I just learned about.

tsutton replied:
The size of the html doc changes by about 10%, and pd.js is about 30% bigger: https://imgur.com/a/7LN6WyY

bakert commented on decksite/templates/rotation.mustache:
```diff
@@ -4,14 +4,12 @@
-            </p>
-        {{/rotation_query}}
+        {{#in_rotation}}
+            <button id="toggle-filters-button" onclick="PD.filter.toggleDisplayFilter()">Show filters</button>
```
If this is identical each time it shows up let’s make it a partial.  You can include partials

{{> likethis}}

bakert commented on shared_web/static/js/pd.js:
```diff
@@ -367,6 +368,164 @@ PD.htmlEscape = function (s) {
+    });
+};
+
+// input url returns a promise to {success: true/false, cardNames: [...], error message: {...}
```
Nit. You are missing a closing curly brace on the end of this comment.

bakert commented on decksite/views/cards.py:
```diff
@@ -9,11 +9,13 @@
         self.show_tournament_toggle = True
         self.tournament_only = tournament_only
+        self.query = query
+        self.cards = cards
```
It’s surprising to me that this view wasn’t setting self.cards already. Without that how were the cards getting vivified for display? How were they being shown in the mustache. Not sure what problems this might cause but it made my spidey sense tingle.

tsutton replied:
This line was unnecessary, self.cards was being set a few lines lower already.

bakert commented on decksite/templates/cardstable.mustache:
```diff
@@ -6,23 +6,33 @@
+                <abbr title="League 5-0 runs">5–0s</abbr>
+            </th>
+            <th class="n" data-breakpoints="xs" data-sortInitialOrder="desc">
+                <abbr title="Number of decks with card in the sideboard"> # SBs</abbr>
```
Is this value already being set or is that in another PR?

tsutton committed:
Improvements to async scryfall filtering in response to comments.

- Change from using HTML id attribute to class and data-* attributes
- Move onclick from HTML to be attached in pd.js
- Add <abbr> tags around symbolic table headers everywhere
- Remove errant reference to sideboard stuff - that's not in this branch.

tsutton committed:
Add missing template file: scryfallfilterwithtoggle.

tsutton committed:
Merge branch 'master' into cards-page-filtering
